const-generic-const-fn-bounds.md

```diff
@@ -0,0 +1,499 @@
+now on the crate author to keep the public API semver compatible, but they can't accidentally fail to
+uphold that obligation by changing private things.
+
+## RPIT (Return position impl trait)
+
+```rust
+const fn foo() -> impl Bar { /* code here */ }
+```
+
+does not allow us to call any methods on the result of a call to `foo`, if we are in a
+const context. It seems like a natural extension to this RFC to allow
+
+```rust
+const fn foo() -> impl const Bar { /* code here */ }
+```
+
+which requires that the function only returns types with `impl const Bar` blocks.
+
+## Specialization
+
+Impl specialization is still unstable. There should be a separate RFC for declaring how
+const impl blocks and specialization interact. For now one may not have both `default`
+and `const` modifiers on `impl` blocks.
+
+## `const` trait methods
+
+This RFC does not touch `trait` methods at all, all traits are defined as they would be defined
+without `const` functions existing. A future extension could allow
+
+```rust
+trait Foo {
+    const fn a() -> i32;
+    fn b() -> i32;
+}
+```
+
+Where all trait impls *must* provide a `const` function for `a`, allowing
+
+```rust
+const fn foo<T: ?const Foo>() -> i32 {
+    T::a()
+}
+```
+
+even though the `?const` modifier explicitly opts out of constness.
+
+The author of this RFC believes this feature to be unnecessary, since one can get the same effect
+by splitting the trait into its const and nonconst parts:
+
+```rust
+trait FooA {
+    fn a() -> i32;
+}
+trait FooB {
+    fn b() -> i32;
+}
+const fn foo<T: FooA + ?const FooB>() -> i32 {
+    T::a()
+}
+```
+
+Impls of the two traits can then decide constness of either impl at their leasure.
+
+### `const` traits
+
+A further extension could be `const trait` declarations, which desugar to all methods being `const`:
+
+```rust
+const trait V {
+    fn foo(C) -> D;
+    fn bar(E) -> F;
+}
+// ...desugars to...
+trait V {
+    const fn foo(C) -> D;
+    const fn bar(E) -> F;
+}
+```
+
+## `?const` modifiers in trait methods
+
+This RFC does not touch `trait` methods at all, all traits are defined as they would be defined
+without `const` functions existing. A future extension could allow
+
+```rust
+trait Foo {
+    fn a<T: ?const Bar>() -> i32;
+}
+```
+
+which does not force `impl const Foo for Type` to now require passing a `T` with an `impl const Bar`
+to the `a` method.
+
+## `const` function pointers
+
+```rust
+const fn foo(f: fn() -> i32) -> i32 {
+    f()
+}
+```
+
+is currently illegal. While we can change the language to allow this feature, two questions make
+themselves known:
+
+1. fn pointers in constants
+
+    ```rust
+    const F: fn() -> i32 = ...;
+    ```
+
+    is already legal in Rust today, even though the `F` doesn't need to be a `const` function.
+
+2. Opt out bounds are ugly
+
+    I don't think it's either intuitive nor readable to write the following
```

Comment:
This is quite funny given that it behaves exactly like the `?const` trait bound. Why do you think it is intuitive or readable there, but not here?

Reply:
> To me it feels like ?const fn() -> i32 is ?(const fn() -> i32) and not (?const) fn() -> i32.

But you don't interpret `?const Trait` as `?(const Trait)` -- why?

Reply:
Not sure, probably because `const fn` is a pattern burned into my mind by now, any variance on it is unusual.

Actually I've started to read `?const fn()` correctly over time, so it's fine, and mostly orthogonal to this RFC I'd think